clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerInterceptor.java

```diff
@@ -0,0 +1,72 @@
+
+
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.TopicPartition;
+
+import java.util.Map;
+
+/**
+ * A plugin interface that allows you to intercept (and possibly mutate) records received by the consumer. A primary use-case
+ * is for third-party components to hook into the consumer applications for custom monitoring, logging, etc.
+ *
+ * <p>
+ * This class will get consumer config properties via <code>configure()</code> method, including clientId assigned
+ * by KafkaConsumer if not specified in the consumer config. The interceptor implementation needs to be aware that it will be
+ * sharing consumer config namespace with other interceptors and serializers, and ensure that there are no conflicts.
+ * <p>
+ * Exceptions thrown by ConsumerInterceptor methods will be caught, logged, but not propagated further. As a result, if
+ * the user configures the interceptor with the wrong key and value type parameters, the consumer will not throw an exception,
+ * just log the errors.
+ * <p>
+ * ConsumerInterceptor callbacks are called from the same thread that invokes {@link org.apache.kafka.clients.consumer.KafkaConsumer#poll(long)}.
+ */
+public interface ConsumerInterceptor<K, V> extends Configurable {
+
+    /**
+     * This is called just before the records are returned by {@link org.apache.kafka.clients.consumer.KafkaConsumer#poll(long)}
+     * <p>
+     * This method is allowed to modify consumer records, in which case the new records will be returned.
+     * Any exception thrown by this method will be caught by the caller, logged, but not propagated to the client.
+     * <p>
+     * Since the consumer may run multiple interceptors, a particular interceptor's onConsume() callback will be called
+     * in the order specified by {@link org.apache.kafka.clients.consumer.ConsumerConfig#INTERCEPTOR_CLASSES_CONFIG}.
+     * The first interceptor in the list gets the consumed records, the following interceptor will be passed the records returned
+     * by the previous interceptor, and so on. Since interceptors are allowed to modify records, interceptors may potentially get
+     * the records already modified by other interceptors. However, building a pipeline of mutable interceptors that depend on the output
+     * of the previous interceptor is discouraged, because of potential side-effects caused by interceptors potentially failing
+     * to modify the record and throwing an exception. If one of the interceptors in the list throws an exception from onConsume(),
+     * the exception is caught, logged, and the next interceptor is called with the records returned by the last successful interceptor
+     * in the list, or otherwise the original consumed records.
+     *
+     * @param records records to be consumed by the client or records returned by the previous interceptors in the list.
+     * @return records that are either modified by the interceptor or same as records passed to this method.
```

Comment:
Are there any limitations regarding the number of records returned? i.e. can an interceptor filter records or generate new records?